### PR TITLE
Add commutative cancellation to preset passmanagers level 2 and 3

### LINKED ISSUE #11911: preset passmanagers should perform gate optimization before routing
### What should we add?

Consider the following circuit:

```python
qc = QuantumCircuit(N)
for kk in range(N - 1):
    qc.cx(kk, N - 1)
qc.x(N - 1)
qc.z(N - 2)
for kk in range(N - 2, -1, -1):
    qc.cx(kk, N - 1)
```

This circuit should collapse down to a pair of X and Z gates via commutation and collapsing pairs of CX gates.  However, the preset passmanagers do not catch this.  Instead, collapsing requires the routing passes to find the correct mapping that allows for later clean up to be successful.  Other SDKs do this, and the performance is markedly better because routing is not required.

### PR DESCRIPTION
**Pull request by sbrandhsn**
### Summary

fixes #11911 

### Details and comments

Added the CommutativeCancellation pass to preset passmanagers level 2 and level 3.

### CODE REVIEW

File: qiskit/circuit/commutation_checker.py
Comment (ElePT) [inline qiskit/circuit/commutation_checker.py]: This is a comment to get to understand the changes, how does this change relate to the addition of the commutative cancellation pass defaults?

Comment (jakelishman) [reply qiskit/circuit/commutation_checker.py]: There's nothing special about `annotated` in particular, so if these guys are causing problems, it might be an indication that there's larger assumptions about the inputs in the `CommutationChecker` right now that need to be examined.

Comment (mtreinish) [reply qiskit/circuit/commutation_checker.py]: Looking at the failure in the first commit, it's because `Operation` doesn't define a `params` field and that's part of the lookup key for the commutation relationship cache: https://github.com/Qiskit/qiskit/blob/main/qiskit/circuit/commutation_checker.py#L97-L98. We probably should just use a `getattr` there and set an empty default. The larger issue we'll have though is for caching is that the lookup key is based on op name and `annotated` is not unique so we'll be caching bogus entries. So if we want to support annotated operations in the commutation analysis/cancellation we'll either have to change the cache key or skip the caching for `AnnotatedOperations`. I think either way fixing `CommutationChecker`'s support for `AnnotatedOperation` could be a standalone followup PR.

Comment (jakelishman) [reply qiskit/circuit/commutation_checker.py]: I think fixing `CommutationChecker`'s logic so it can run on abstract circuits is probably important to do before we merge this PR, since we'd be moving the pass from its current position of running on physical circuits (where it can typically assume it'll only see `Instruction` subclasses) to one where it'll definitely see abstract `Operation` classes.

That said, I agree those fixes should be in a separate PR.

Comment (brandhsn) [reply qiskit/circuit/commutation_checker.py]: I agree with these points. Thanks for raising them! :-) Is there a reliable way to check whether an operation is described by a (unitary) matrix (and also to retrieve that matrix), i.e. can we just do `hasattr(obj, "to_matrix")`? Can retrieving the matrix have a large runtime overhead for an arbitrary `Operation`? If so we might not be willing to do that for commutation cancellation but maybe the user should decide that.

Should the name of an annotated operation always be `"annotated"+base_op.name` or should we handle this in the commutation checker? Maybe skipping caching for an abstract operation makes sense...

Comment (mtreinish) [reply qiskit/circuit/commutation_checker.py]: > I agree with these points. Thanks for raising them! :-) Is there a reliable way to check whether an operation is described by a (unitary) matrix (and also to retrieve that matrix), i.e. can we just do hasattr(obj, "to_matrix")? Can retrieving the matrix have a large runtime overhead for an arbitrary Operation? If so we might not be willing to do that for commutation cancellation but maybe the user should decide that.

Yeah we can just do `hasattr(obj, "to_matrix")`, in the general case it should be fast for most gates. This is actually what the `Operator` class does already: https://github.com/Qiskit/qiskit/blob/main/qiskit/quantum_info/operators/operator.py#L105 But there is nothing in practice blocking a particular gate object from implementing an expensive function, but I personally wouldn't worry about that to try and guard against. 

> Should the name of an annotated operation always be "annotated"+base_op.name or should we handle this in the commutation checker? Maybe skipping caching for an abstract operation makes sense...

It would need to be `annotate+base_op.name+modifiers` to be a unique look up key. But I would maybe argue against trying to cache annotate operations, I find it unlikely that in practice we'll have a large number of repeated instances of the same exact annotated operation in a circuit.  I think just doing the matrix multiplication whenever we encounter an annotate op makes the most sense.

Comment (sbrandhsn) [reply qiskit/circuit/commutation_checker.py]: I hope I addressed these points in PR #11948

File: releasenotes/notes/commutative-cancellation-preset-passmanager-c137ce516a10eae5.yaml
Comment (mtreinish) [inline releasenotes/notes/commutative-cancellation-preset-passmanager-c137ce516a10eae5.yaml]:
```suggestion
    Added the :class`.CommutativeCancellation` pass to the ``init`` stage of the preset pass managers for
    optimization levels 2 and 3. This enables the preset pass managers to cancel additional logical
    operations at the beginning of the compilation pipeline.
    
```

File: releasenotes/notes/commutative-cancellation-preset-passmanager-c137ce516a10eae5.yaml
Comment (mtreinish) [inline releasenotes/notes/commutative-cancellation-preset-passmanager-c137ce516a10eae5.yaml]: Oops my bad I had a typo in my suggestion:

```suggestion
    Added the :class:`.CommutativeCancellation` pass to the ``init`` stage of the preset pass managers for
```

**Review by mtreinish — APPROVED** [review]

**Pull request merged**
